codec/tools/strings.go

```diff
@@ -0,0 +1,130 @@
+package codectools
+
+import (
+	"bytes"
+	"io"
+	"unicode/utf8"
+)
+
+var hex = "0123456789abcdef"
+
+// EscapeString is as per WriteEscapedString, but returns the result rather than pushing it into an io.Writer.
+func EscapeString(s string) string {
+	var buf bytes.Buffer
+	WriteEscapedString(s, &buf)
+	return buf.String()
+}
+
+func QuoteEscapeString(s string) string {
+	var buf bytes.Buffer
+	WriteQuotedEscapedString(s, &buf)
+	return buf.String()
+}
+
+func WriteQuotedEscapedString(s string, w io.Writer) error {
+	if _, err := w.Write([]byte{'"'}); err != nil {
+		return err
+	}
+	if err := WriteEscapedString(s, w); err != nil {
+		return err
+	}
+	if _, err := w.Write([]byte{'"'}); err != nil {
+		return err
+	}
+	return nil
+}
+
+// WriteEscapedString emits a string, surrounded by quotation marks, that's escaped in what's probably "DWIM"-compliant.
+//
+// More specifically:
+//
+//   - The output always begins and ends in `"`.
+//   - Any `"` marks in the string are escaped as `\"`.
+//   - The common escape sequences for familiar unprintables is used: `\r`, `\n`, `\t`, etc.
+//   - Unprintables are encoded as `\x__`, where `__` is replaced by the two lowercase hex characters for that byte.
+//   - UTF-8 runes will be printed as they are (no escaping; e.g., å is å).
+//   - UTF invalid sequences will be encoded as if they were unprintables.
+//   - Any `\` characters in the string are escaped as `\\`.
+//
+// The resulting output will be a single line, and can be parsed back to the exact same string as the original.
+// The resulting output is parsable as a JSON string if it does not include any `\x__` sequences,
+// or if your json parser supports hex escapes.
+//
+func WriteEscapedString(s string, w io.Writer) error {
```

Comment:
I'm really, really surprised that you went for this instead of reusing Go's or JSON's string quoting. It's an extra 100 lines of very tricky code that we would now have to maintain for little benefit :)

Reply:
I wrote a longer comment on that in the main thread :)  Character encoding is a beastly topic.